src/mip/coefficient_strengthening.jl

```diff
@@ -0,0 +1,169 @@
+
+function upperbound_strengthening!(ps::PresolveData{T}, i::Int, j_index::Int, j::Int, max_act) where {T}
+    # perform coef strengthening for one constraints of the from a'x <= u
+    row = ps.pb0.arows[i]
+    a = row.nzval
+    u = ps.urow[i]
+    if a[j_index] > 0
+        d = u - max_act - a[j_index]*(ps.ucol[j]-1)
+        if a[j_index] >= d > 0
+            a[j_index] = a[j_index] - d
+            ps.urow[i] = u - d*ps.ucol[j]
+        end
+    elseif a[j_index] < 0
+        d = u - max_act - a[j_index]*(ps.lcol[j]+1)
+        if -a[j_index] >= d > 0
+            a[j_index] = a[j_index] + d
+            ps.urow[i] = u + d*ps.lcol[j]
+        end
+    end
+    return T(a[j_index])
+end
+
+function lowerbound_strengthening!(ps::PresolveData{T}, i::Int, j_index::Int, j::Int, min_act) where {T}
+    # perform coef strengthening for one constraints of the from l < = a'x
+    row = ps.pb0.arows[i]
+    a = row.nzval
+    l = ps.lrow[i]
+    if a[j_index] > 0
+        d = -l + min_act + a[j_index]*(ps.lcol[j]+1)
+        if a[j_index] >= d > 0
+            a[j_index] = a[j_index] - d
+            ps.lrow[i] = l - d*ps.lcol[j]
+        end
+    elseif a[j_index] < 0
+        d = -l - min_act + a[j_index]*(ps.ucol[j]-1)
+        if -a[j_index] >= d > 0
+            a[j_index] = a[j_index] + d
+            ps.urow[i] = u + d*ps.ucol[j]
+        end
+    end
+    return T(a[j_index])
+end
+
+function zero_coefficient_strengthening!(ps::PresolveData{T}) where {T}
+    # perform coefficient stregthening but if there is a coefficient is reduced to 0
+    # it is still kept in the ps.pb0.arows and ps.pb0.acols
+
+    # keep track of index for each var fo update ps.acols
+    # use this to find which index of ps.pb0.acols[i].nzval to update
+    i_index = zeros(Int, ps.pb0.nvar)
+
+    for i in 1:ps.pb0.ncon
+
+        lrow = ps.lrow[i]
+        urow = ps.urow[i]
+        if lrow > -Inf && urow < Inf #skip ranged constraints
+            continue
+        end
+
+        row = ps.pb0.arows[i]
+        if all(ps.var_types[row.nzind] .== CONTINUOUS) # at least 1 integer
+            continue
+        end
+
+        sup = maximal_activity(row, ps.ucol, ps.lcol)
+        inf = minimal_activity(row, ps.ucol, ps.lcol)
+
+        j_index = 0 # keep track of index of variable j in row.nzind & row.nzval
+        for j in row.nzind
+            j_index += 1
+            i_index[j] += 1
+            if ps.var_types[j] == CONTINUOUS || !ps.colflag[j]
+                continue
+            else
+                coef = row.nzval[j_index]
+
+                if urow < Inf
+                    if coef > 0
+                        max_act = sup - coef * ps.ucol[j] # maximal activity of every variables except j
+                    else
+                        max_act = sup - coef * ps.lcol[j]
+                    end
+                    new_coef = upperbound_strengthening!(ps, i, j_index, j, max_act)
+                    ps.pb0.acols[j].nzval[i_index[j]] = new_coef
```

Comment:
I think it would be more consistent to have the code updating row `i` and column `j` in the same place: either both here, or both within `upperbound_strengthening!`.